cpp/src/request_am.cpp

```diff
@@ -230,59 +233,95 @@ ucs_status_t RequestAm::recvCallback(void* arg,
   }
 }
 
-std::shared_ptr<Buffer> RequestAm::getRecvBuffer() { return _buffer; }
+std::shared_ptr<Buffer> RequestAm::getRecvBuffer()
+{
+  return std::visit(data::dispatch{
+                      [](data::AmReceive amReceive) { return amReceive._buffer; },
+                      [](auto arg) {
+                        throw std::runtime_error("Unreachable");
+                        return decltype(data::AmReceive::_buffer){nullptr};
+                      },
+                    },
+                    _requestData);
+}
 
 void RequestAm::request()
 {
-  static const ucp_tag_t tagMask = -1;
-
-  ucp_request_param_t param = {.op_attr_mask = UCP_OP_ATTR_FIELD_CALLBACK |
-                                               UCP_OP_ATTR_FIELD_FLAGS |
-                                               UCP_OP_ATTR_FIELD_USER_DATA,
-                               .flags     = UCP_AM_SEND_FLAG_REPLY,
-                               .datatype  = ucp_dt_make_contig(1),
-                               .user_data = this};
-
-  _sendHeader = _delayedSubmission->_memoryType;
-
-  if (_delayedSubmission->_send) {
-    param.cb.send = _amSendCallback;
-    void* request = ucp_am_send_nbx(_endpoint->getHandle(),
-                                    0,
-                                    &_sendHeader,
-                                    sizeof(_sendHeader),
-                                    _delayedSubmission->_buffer,
-                                    _delayedSubmission->_length,
-                                    &param);
-
-    std::lock_guard<std::recursive_mutex> lock(_mutex);
-    _request = request;
-  } else {
-    throw ucxx::UnsupportedError(
-      "Receiving active messages must be handled by the worker's callback");
-  }
+  std::visit(data::dispatch{
+               [this](const data::AmSend& amSend) {
```

Comment:
Here we must get `data::AmSend` by reference because we pass `amSend._memoryType` by reference to `ucp_am_send_nbx`. I'm not sure what should be the safer and cleaner way, but perhaps we should visit objects always by reference instead of by value as is done elsewhere (one notable exception is https://github.com/rapidsai/ucxx/pull/121/files#diff-637c52f56b253c39f497412ac96ffbb95ba303c20f6528f9f66e902601a1338dR21) when we need to modify the object in ) and make sure they're `const` where we don't need to modify it?

Reply:
> Here we must get `data::AmSend` by reference because we pass `amSend._memoryType` by reference to `ucp_am_send_nbx` which may not complete sychronously. I'm not sure what should be the safer and cleaner way, but perhaps we should visit objects always by reference instead of by value as is done elsewhere (one notable exception is https://github.com/rapidsai/ucxx/pull/121/files#diff-637c52f56b253c39f497412ac96ffbb95ba303c20f6528f9f66e902601a1338dR21) when we need to modify the object in ) and make sure they're `const` where we don't need to modify it?

Reply:
This is because we need the `header` buffer to live for as long as the lifetime of the request returned by `ucp_am_send_nbx` I think?

If we set `.flags = UCP_AM_SEND_FLAG_REPLY | UCP_AM_SEND_FLAG_COPY_HEADER` then UCX will memcpy the header and it will only need to be live until the call to `ucp_am_send_nbx` completes.

That might be a cleaner way to handle things.

Reply:
Yes, thanks for pointing that out. I didn't check whether there was a flag to do that. In the future that might not be the case anymore though, I made the choice of `ucs_memory_type_t` to act as header here for simplicity, ideally we would switch to some type that can be customized by the user and not anymore limit active messages to just sending/receiving a single hardcoded buffer type and give the freedom for the user to transfer complex types too. One "user" that could benefit from that is multi-buffer transfers, currently it's only supported by tag but I think we could make a "multi-buffer type" that we could send and receive via a single active message UCX call (I haven't thought too much about it though so complications may exist that make my idea not applicable).

Reply:
Adding `UCP_AM_SEND_FLAG_COPY_HEADER ` in https://github.com/rapidsai/ucxx/pull/121/commits/fc63a0ed7d381f03b1431208b196cd0709551d26 .